src/enclave/rpcendpoint.h

```diff
@@ -34,31 +35,113 @@ namespace enclave
       session_id(session_id)
     {}
 
+    std::optional<jsonrpc::Pack> detect_pack(const std::vector<uint8_t>& input)
+    {
+      if (input.empty())
+        return std::nullopt;
+
+      if (input[0] == '{')
+        return jsonrpc::Pack::Text;
+      else
+        return jsonrpc::Pack::MsgPack;
+    }
+
+    std::pair<bool, nlohmann::json> unpack_json(
+      const std::vector<uint8_t>& input, jsonrpc::Pack pack)
+    {
+      nlohmann::json rpc;
+      try
+      {
+        rpc = jsonrpc::unpack(input, pack);
+        if (!rpc.is_object())
+          return jsonrpc::error(
+            jsonrpc::StandardErrorCodes::INVALID_REQUEST,
+            fmt::format("RPC payload is a not a valid object: {}", rpc.dump()));
+      }
+      catch (const std::exception& e)
+      {
+        return jsonrpc::error(
+          jsonrpc::StandardErrorCodes::INVALID_REQUEST,
+          fmt::format("Exception during unpack: {}", e.what()));
+      }
+
+      return {true, rpc};
+    }
+
+    std::optional<std::string> get_method(const nlohmann::json& j)
+    {
+      if (j.find(jsonrpc::SIG) != j.end())
+      {
+        return j.at(jsonrpc::REQ).at(jsonrpc::METHOD).get<std::string>();
+      }
+      else
+      {
+        return j.at(jsonrpc::METHOD).get<std::string>();
+      }
+    }
+
+    std::string get_actor(const std::string& method)
+    {
+      return method.substr(0, method.find_last_of('/'));
+    }
+
     bool handle_data(const std::vector<uint8_t>& data)
     {
-      if (!handler)
+      LOG_TRACE_FMT("Entered handle_data {} {}", data.size(), data.empty());
+      auto pack = detect_pack(data);
+      if (!pack.has_value())
+      {
+        LOG_TRACE_FMT("NO PACK");
+        send(jsonrpc::pack(
+          jsonrpc::error_response(
+            0, jsonrpc::StandardErrorCodes::INVALID_REQUEST, "Empty request."),
+          jsonrpc::Pack::Text));
+        return true;
+      }
+
+      LOG_TRACE_FMT("Detected");
+      auto [deserialised, rpc] = unpack_json(data, pack.value());
+
+      if (!deserialised)
       {
-        // The hostname indicates the rpc class.
-        auto host = hostname();
+        send(jsonrpc::pack(rpc, pack.value()));
```

Comment:
does this need to be re-packed, can we not just send `data`?

Reply:
No, this is the response "couldn't deserialised", the naming is slightly confusing, but it isn't the original RPC.